### ISSUE TITLE
Type substitution info is lost when emiting nested generic struct field

### ISSUE THREAD

**Issue by csyonghe** [shader-slang/slang#292]
```
interface IInterpolant
{
    float4 getColor();
}
__generic_param TVertInterpolant : IInterpolant;
struct Impl : IInterpolant
{
    float4 color;
    float4 getColor() { return color; }
};

struct VertexStageOutput
{
    float4 sv_position : SV_Position;
    TInterpolant interpolants;
};

VertexStageOutput vsMain()
{
   ...
}
```

When emiting final HLSL from IR (with type arguments = {Impl}), `VertexStageOutput` will get the following definition:
```
struct VertexStageOutput
{
    float4 sv_position : SV_Position;
    TInterpolant interpolants;
};
```
This is wrong, we actual want
```
struct VertexStageOutput
{
    float4 sv_position : SV_Position;
    Impl interpolants;
};
```

The reason is that when `ensureStructDecl()` is called on `DeclRefType<VertexStageOutput, subst = {GlobalGenericTypeSubstitution(TInterpolant->Impl)}>`, it iterates through the member fields of `VertexStageOutput` and recurse on `field->type`. But `field->type` does not include the required substitution.

We need a mechanism to pass the substitution down to nested child types.

**Comment by tfoleyNV** [comment]
This should be a simple fix. You can do: `getMembersOfType<StructField>(aggTypeDeclRef)` to get decl refs to the fields, and then `GetType(fieldDeclRef)` to get a substituted type for the field. The support is all there and it is just that the code in question is failing to use if.

**Comment by tfoleyNV** [comment]
Okay, never mind. My previous comment would have been accurate for the case of an ordinary generic `struct` type, but doesn't account for global generic parameters.

It seems like the only real fix is to attach the global generic substitutions to every type or `DeclRef` that *might* need them, possibly as part of generating the target-specific IR, and then get those substituted types to propagate through the IR and to emit time.

**Comment by csyonghe** [comment]
Fixed in #293
